### ISSUE TITLE
filter.d/sendmail-reject.conf

### ISSUE THREAD

**Issue by cbjacques** [fail2ban/fail2ban#1699]
I just upgraded fail2ban from 0.8.11-1 to 0.9.4.
The new sendmail-reject.conf filter does not catch certain sendmail errors.  Specifically, I get a lot of "knocking" attempts.  When this happens, I get an error in my mail.log that looks like this:
```
Feb 20 06:25:36 [my server IP] sm-mta[22765]: v1KBPYxg022765: [198.240.124.169] did not issue MAIL/EXPN/VRFY/ETRN during connection to MTA-v4
```
Unfortunately, the new sendmail-reject.conf regex doesn't catch this.  The regex it ships with is:
```
^%(__prefix_line)s\w{14}: (\S+ )?\[<HOST>\]: ((?i)expn|vrfy) \S+ \[rejected\]$
```

I tried fiddling with it to make it work, but failed.  Eventually, I just cut an paste the old regex from 0.8 in:

```
 \[<HOST>\] did not issue MAIL/EXPN/VRFY/ETRN during connection to (MSP|MTA)-v\d$
```

That solved my problem.  Honestly, I'd fix it but I'm not very good at python.

Here's my configuration:
fail2ban 0.9.4
ubuntu 14.04.5 LTS "trusty"
sendmail 8.14.4
python 2.7.6

**Comment by sebres** [comment]
> I just cut an paste the old regex from 0.8 in

This supposedly working regex was never part of the stock fail2ban filter `sendmail-reject`.
See [history of sendmail-reject.conf](https://github.com/fail2ban/fail2ban/commits/master/config/filter.d/sendmail-reject.conf)

Either it is a maintainer issue (some customized filter for your distro) or I'm missing something...
Please provide more log-lines if they available (I need larger log-excerpt, because it's multi-line filter).

BTW. Your provided supposedly working regex can be vulnerable (by injecting on user info)

**Comment by cbjacques** [comment]
My mail.log file contains logs for dovecot and spamassassin as well, so it's crowded.  The errors that sendmail-reject is failing to catch contain "EXPN".  So I grep'd the mail.log +/- 5 lines from sm-mta matching EXPN.  That file is attached.

I thought the regex I used was from an earlier build, but it must have been _reported_ as from an earlier build.  Sorry for the confusion.

[sm-mta-log.txt](https://github.com/fail2ban/fail2ban/files/834419/sm-mta-log.txt)

**Comment by sebres** [comment]
Thx for the log-excerpt.

It looks like not an authentication errors (so too aggressive for normal usage).
I'll make a parameter `mode` (set on `aggressive`, default `normal`) for the 0.10 branch to cover this kind of failures (to let the users decide about).

For the 0.9th branch you can use this one in your `sendmail-reject.local` (please don't modify the stock .conf files):
```
[Definition]
failregex = %(known/failregex)s
            ^%(__prefix_line)s\w{14}: (?:\S+ )?\[<HOST>\] did not issue (?:[A-Z]{4}[/ ]?)*during connection to MTA(?:-\w+)?$
```

BTW I found in your excerpt another possible failure (rejected with reason "Mailbox disabled for this recipient"). It look like really auth-failure (possibly I'll acquire it in the stock filter).

If interested you can use this one (combines both):
```
[Definition]
failregex = %(known/failregex)s
            ^%(__prefix_line)s\w{14}: ruleset=check_rcpt, arg1=(?P<email><\S+@\S+>), relay=(\S+ )?\[<HOST>\]( \(may be forged\))?, reject=550 5\.2\.1 (?P=email)\.\.\. Mailbox disabled for this recipient$
            ^%(__prefix_line)s\w{14}: (?:\S+ )?\[<HOST>\] did not issue (?:[A-Z]{4}[/ ]?)*during connection to MTA(?:-\w+)?$
```

**Comment by sebres** [comment]
Small amend (missing MSP-mode and "may be forged"):
```
[Definition]
failregex = %(known/failregex)s
            ^%(__prefix_line)s\w{14}: ruleset=check_rcpt, arg1=(?P<email><\S+@\S+>), relay=(\S+ )?\[<HOST>\](?: \(may be forged\))?, reject=550 5\.2\.1 (?P=email)\.\.\. Mailbox disabled for this recipient$
            ^%(__prefix_line)s\w{14}: (?:\S+ )?\[<HOST>\](?: \(may be forged\))? did not issue (?:[A-Z]{4}[/ ]?)+during connection to M(?:TA|SP)(?:-\w+)?$
```

**Comment by cbjacques** [comment]
"Mailbox disabled" is something we do on our end.  It does not represent a hacking attempt.  Fail2ban does not need to look at it.  Valid hosts could transmit mail to a disabled mailbox.

"did not issue MAIL/EXPN/VRFY/ETRN" is frequently a hacking attempt.  It can also result from poorly configured mail agents.   Indeed, almost all of the tagged hosts are from far away (like China) and have no business connecting to our mail server.

**Comment by sebres** [comment]
> Valid hosts could transmit mail to a disabled mailbox.

Hmm? Could you provide an example where it would be possible (resp. where it may be desirable)

>  "did not issue" ... It can also result from poorly configured mail agents

Therefore I said "it's too aggressive".

**Comment by cbjacques** [comment]
For example, we used to have an email account clp@mydomain.com.  It got way too much spam, so we shut it off.  Any email directed to that address will receive the "disabled mailbox".  You could, for example, send mail from Google to that email address and it would be perfectly valid.   We wouldn't want to ban valid email hosts from sending mail to a disabled address.

**Comment by sebres** [comment]
That would be right if google would use a sendmail way to mail you (what it does not do). 
I thought we speak still about sendmail-reject filter?

But in the meantime I think to add this optional also.

**Comment by cbjacques** [comment]
Just to be clear, this error does not indicate some body is trying to _log in_ to a disabled account.  It indicates someone tried to send mail to a disabled account.

Google does use sendmail.  I just emailed myself from Google.  I've attached a sanitzed log file from the event.

If I send mail from google to the disabled account, that will also get logged.  If fail2ban flags it, then google will get disabled; nobody would be able to send me mail from a google account regardless of who they were sending it to.

I do not think fail2ban should be looking for attempts to send mail to a disabled account. 

[gsm.txt](https://github.com/fail2ban/fail2ban/files/835000/gsm.txt)
